ndk/src/font.rs

```diff
@@ -0,0 +1,457 @@
+            FontWeight::BLACK => f.write_str("Black (Heavy)"),
+            _ => writeln!(f, "{}", self.0),
+        }
+    }
+}
+
+/// The error type returned when an invalid font weight value is passed.
+#[derive(Debug)]
+pub struct FontWeightValueError(());
+
+impl fmt::Display for FontWeightValueError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str("font weight must be positive and less than or equal to 1000")
+    }
+}
+
+impl std::error::Error for FontWeightValueError {}
+
+impl TryFrom<u16> for FontWeight {
+    type Error = FontWeightValueError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        FontWeight::new(value)
+    }
+}
+
+/// A 4-byte integer representing an OpenType axis tag.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct AxisTag(u32);
+
+impl AxisTag {
+    pub const fn from_be(value: u32) -> Result<Self, AxisTagValueError> {
+        // Each byte in a tag must be in the range 0x20 to 0x7E. A space character cannot be
+        // followed by a non-space character. A tag must have one to four non-space characters.
+        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#data-types for details.
+
+        let bytes = value.to_be_bytes();
+
+        // Each byte in a tag must be in the range 0x20 to 0x7E.
+        macro_rules! check_byte_range {
+            ($($e:expr)+) => {
+                $(
+                    if !(bytes[$e] as char).is_ascii_graphic() && bytes[$e] != b' ' {
+                        return Err(AxisTagValueError::InvalidCharacter);
+                    }
+                )+
+            };
+        }
+        check_byte_range!(0 1 2 3);
+
+        if bytes[0] == b' ' {
+            return Err(
+                if bytes[1] == b' ' && bytes[2] == b' ' && bytes[3] == b' ' {
+                    // A tag must have one to four non-space characters.
+                    AxisTagValueError::EmptyTag
+                } else {
+                    // A space character cannot be followed by a non-space character.
+                    AxisTagValueError::InvalidSpacePadding
+                },
+            );
+        }
+
+        macro_rules! check_if_valid {
+            ($e:expr ; $($f:expr)+) => {
+                if bytes[$e] == b' ' {
+                    return if true $(&& bytes[$f] == b' ')+ {
+                        Ok(Self(value))
+                    } else {
+                        // A space character cannot be followed by a non-space character.
+                        Err(AxisTagValueError::InvalidSpacePadding)
+                    };
+                }
+            };
+        }
+
+        check_if_valid!(1; 2 3);
+        check_if_valid!(2; 3);
+
+        // Whether or not bytes[3] is b' ', value is a valid axis tag.
+        Ok(Self(value))
+    }
+
+    pub const fn from_be_bytes(value: [u8; 4]) -> Result<Self, AxisTagValueError> {
+        Self::from_be(u32::from_be_bytes(value))
```

Comment:
How about doing it the opposite way around, by having all the logic inside `from_be_bytes()` and make `from_be(u32)` call `from_be_bytes(value.to_be_bytes())`?